cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -397,9 +397,52 @@ public void apply(MetadataDataset input) throws Exception {
     List<Map.Entry<Id.NamespacedId, Integer>> resultList = new ArrayList<>(weightedResults.entrySet());
     Collections.sort(resultList, SEARCH_RESULT_DESC_SCORE_COMPARATOR);
 
+    // Fetch metadata for entities in the result list
+    // Note: since the fetch is happening in a different transaction, the metadata for entities may have been
+    // removed. It is okay not to have metadata for some results in case this happens.
+    Map<Id.NamespacedId, Metadata> systemMetadata = fetchMetadata(weightedResults.keySet(), MetadataScope.SYSTEM);
+    Map<Id.NamespacedId, Metadata> userMetadata = fetchMetadata(weightedResults.keySet(), MetadataScope.USER);
+
+    return addMetadataToResults(resultList, systemMetadata, userMetadata);
+  }
+
+  private Map<Id.NamespacedId, Metadata> fetchMetadata(final Set<Id.NamespacedId> entityIds, MetadataScope scope) {
+    Set<Metadata> metadataSet =
+      execute(new TransactionExecutor.Function<MetadataDataset, Set<Metadata>>() {
+        @Override
+        public Set<Metadata> apply(MetadataDataset input) throws Exception {
+          return input.getMetadata(entityIds);
+        }
+      }, scope);
+    Map<Id.NamespacedId, Metadata> metadataMap = new HashMap<>();
+    for (Metadata m : metadataSet) {
+      metadataMap.put(m.getEntityId(), m);
+    }
+    return metadataMap;
+  }
+
+  Set<MetadataSearchResultRecord> addMetadataToResults(List<Map.Entry<Id.NamespacedId, Integer>> results,
+                                                       Map<Id.NamespacedId, Metadata> systemMetadata,
+                                                       Map<Id.NamespacedId, Metadata> userMetadata) {
     Set<MetadataSearchResultRecord> result = new LinkedHashSet<>();
-    for (Map.Entry<Id.NamespacedId, Integer> entry : resultList) {
-      result.add(new MetadataSearchResultRecord(entry.getKey()));
+    for (Map.Entry<Id.NamespacedId, Integer> entry : results) {
+      ImmutableMap.Builder<MetadataScope, co.cask.cdap.proto.metadata.Metadata> builder = ImmutableMap.builder();
+      // Add system metadata
+      Metadata metadata = systemMetadata.get(entry.getKey());
+      if (metadata != null) {
+        builder.put(MetadataScope.SYSTEM,
+                    new co.cask.cdap.proto.metadata.Metadata(metadata.getProperties(), metadata.getTags()));
```

Comment:
can the internal object be superseeded with the one in ``cdap-proto``? Or are there any differences between them?